### ISSUE TITLE
include Subject Alternative Name extension in ESGF host certificates

### ISSUE THREAD

**Issue by alaniwi** [ESGF/esgf-installer#281]
Chrome and Firefox seem to insist that a host certificate uses the Subject Alternative Name extension to specify the hostname, rather than the CN, even if it only has one DNS name.  The certificates in ESGF do not have this.  Here is an example of what it should look like (this certificate from University of Oxford):
```
        X509v3 extensions:
[...]
            X509v3 Subject Alternative Name: 
                DNS:aurochs-web-web-154.nsms.ox.ac.uk, DNS:aurochs-web-web-155.nsms.ox.ac.uk, DNS:aurochs.nsms.ox.ac.uk, DNS:ox.ac.uk, DNS:oxfordtoday.ox.ac.uk, DNS:www.clarendon.ox.ac.uk, DNS:www.ox.ac.uk, DNS:www.oxfordtoday.ox.ac.uk

```
As an example, I had a host certificate for HTTPS signed by IPSL, and then tried importing into the browser the IPSL CA certificate, but the browser still did not accept the host validation unless I added an exception. By contrast, `wget` accepted it without needing `-no-check-certificate` if I pointed at the IPSL certificate by using `--ca-certificate=0597e90c.crt`. I believe this is because `wget` is using the `CN` entry when validating.

This could be an issue even for commercially signed certificates, if the certificate request is generated by the ESGF software before sending it to the commercial CA for signing. In fact with the commercial certificates which we use for our operational nodes, the company generates the CSR for us, and then sends us the signed certificate and the private key -- and the certificate they gave us has the "Subject Alternative Name" extension.  However, it is possible that others may wish to generate their own CSRs, particularly if they want to keep their private key completely private.

Unfortunately I don't know how to generate a CSR with this extension. I tried some instructions I found, but got an error about `X509 V3 routines:DO_EXT_NCONF:unknown extension name:v3_conf.c`

----
_Incidentally, what I actually sent to IPSL for signing was the local CA certificate rather than the host certificate, so I signed the host key myself using the local CA, and the certificate chain in this case looks like:_
```
 0 s:/O=ESGF/OU=ESGF.ORG/CN=esgf-test-install.ceda.ac.uk
   i:/O=ESGF/OU=ESGF.ORG/CN=esgf-test-install.ceda.ac.uk-CA

 1 s:/O=ESGF/OU=ESGF.ORG/CN=esgf-test-install.ceda.ac.uk-CA
   i:/O=ESGF/OU=ESGF.ORG/OU=simpleca.ipsl.upmc.fr/CN=IPSL Simple CA

 2 s:/O=ESGF/OU=ESGF.ORG/OU=simpleca.ipsl.upmc.fr/CN=IPSL Simple CA
   i:/O=ESGF/OU=ESGF.ORG/OU=simpleca.ipsl.upmc.fr/CN=IPSL Simple CA
```
_rather than having the host key directly signed by IPSL, as is the more usual situation in ESGF. I don't *think* this is relevant to the problem, particularly as `wget` is happy with it, but I mention it just in case._

**Comment by alaniwi** [comment]
https://www.chromestatus.com/feature/4981025180483584

**Comment by alaniwi** [comment]
Thanks for letting me know. Two comments:

* Even for test nodes (where we presumably don't want to insist on commercially signed certs), it would be helpful to have certificates in the format that browsers now expect. At least Firefox lets me add an exception, but Chrome does not even do that (it just says "esgf-test-install.ceda.ac.uk doesn't adhere to security standards. ERR_SSL_SERVER_CERT_BAD_FORMAT"). 

* In any case, my third paragraph above (starting "This could be an issue even for commercially signed certificates...") still applies if we generate CSRs that people might use to obtain the commercially signed certs.

**Comment by pchengi** [comment]
I've made changes to the NSC SimpleCA. I made tests and found that we don't need to have the SAN in the request at all; it can be embedded in the signed cert directly. For our ESGF CAs, it's acceptable to have a SAN which is the same value as the hostname, so no need of asking an additional question or configuring anything extra. I've also made changes to the ESGF CA scripts, and informed Lukasz and Sebastien; I've also reissued certs to all nodes using NSC SimpleCA certs, with the SAN extension enabled. If you have a cert issued by ANL or IPSL, contact them and get your certificate reissued.
